---
title: "Gen_model_gene_Importance_modification"
output: html_document
date: "2025-06-23"
note: "This script modifies the feature importance files to remove the aneuploidy feature cis-genes from the importance output and save the modified files."
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = "D:/projects/HetIN_signatures/")
```


Load the packages
```{r}
library(dplyr)
library(tidyverse)
library(ggplot2)
library(rtracklayer)
library(GenomicRanges)
library(patchwork)
library(ggnewscale)
```

Load the data and set up the environment
```{r}

# folder with the feature importance files
feat_imp_folder <- "data/Model_output/Results/Feature_importance/PANCAN/Base_feat_imp"
base_feat_imp_plot <- "data/Model_output/Plots/Feature_importance/PANCAN/Base_feat_imp"

# Load the gene annotations 
gene_annotations <- read.csv("data/secondary_data/gene_seqnames_nnotation_df.csv", stringsAsFactors = FALSE, row.names = NULL)

# Load the gene_set genes
gene_set_genes <- read.csv("data/secondary_data/Gene_set.csv", stringsAsFactors = FALSE)
gene_set <- colnames(gene_set_genes)

# reduce the gene annotations to the genes that are in the set of interest
gene_annotations <- gene_annotations %>%
  filter(gene_name %in% gene_set) %>% 
  mutate(
    chr_arm = paste0(seqnames, arm), # combine the chromosome and arm into new column
  )

# Set the order for the response features
response_features_order <- zipup(
  paste0(1:22, "p"), 
  paste0(1:22, "q")
) %>% 
  unlist() 
```

Loop through modify and save the feature importance files
```{r}
# Define the folders for the feature importance files 
complete_imp_folder <- file.path(feat_imp_folder, "complete")
arm_lvl_filt_imp_folder <- file.path(feat_imp_folder, "arm_lvl_filtered")
chrm_lvl_filt_imp_folder <- file.path(feat_imp_folder, "chrm_lvl_filtered")
ori_imp_folder <- file.path(feat_imp_folder, "original")

# Create the complete and filtered folders if they do not exist
if (!dir.exists(complete_imp_folder)) {
  dir.create(complete_imp_folder, recursive = TRUE)
}
if (!dir.exists(arm_lvl_filt_imp_folder)) {
  dir.create(arm_lvl_filt_imp_folder, recursive = TRUE)
}
if (!dir.exists(chrm_lvl_filt_imp_folder)) {
  dir.create(chrm_lvl_filt_imp_folder, recursive = TRUE)
}

# Initialize empty dataframes to store the results
complete_cancer_type_imp_df <- data.frame()
arm_lvl_filtered_cancer_type_imp_df <- data.frame()
chrm_lvl_filtered_cancer_type_imp_df <- data.frame()

# initialize sets to store the features by type
hrd_set <- c("ai1", "lst1", "loh_hrd")
peri_set <- c()
aneu_set <- c()


# Loop through each feature importance files in the folder
for (feature_file in list.files(ori_imp_folder, full.names = TRUE)) {
  # cat("\n\n", feature_file)
  
  # Get the feature from the name
  file_name <- basename(feature_file)
  # cat("\n File name: ", file_name)
  
  name_split <- str_split(file_name, "_")[[1]]
  # cat("\n\n Name split: ", paste(name_split, collapse = ", "))
  first_part <- name_split[1]
  # cat("\n First part: ", first_part)
  
  # Define the target feature based on the file name
  if (first_part == "loh") {
    target_feat <- paste(name_split[1:2], collapse = "_")
  } 
  else if (first_part == "peri") {
    if (length(name_split) > 4) {
      target_feat <- paste(name_split[1:3], collapse = "_")
    } else {
      target_feat <- paste(name_split[1:2], collapse = "_")
    }
    peri_set <- append(peri_set, target_feat) # add to the peri_set if peri feature
  }
  else {
    target_feat <- first_part
    chr <- substr(target_feat, 1, nchar(target_feat) - 1) 
    cat("\n Chromosome: ", chr)
  }
  cat("\n Target_feat: ", target_feat)
    
  # Read the CSV file
  feature_df <- read.csv(feature_file, stringsAsFactors = FALSE, row.names = NULL) %>% 
    select(-X)
  
  # Join the feature importance dataframe with the gene annotations
  feat_joind_df <- feature_df %>%
    left_join(
      gene_annotations %>% 
        select(gene_name,seqnames, arm), 
      by = c("Feature" = "gene_name")
      ) %>% dplyr::rename(Chr = seqnames, Arm = arm) %>%
    drop_na(Chr) %>% # drop rows with NA in Chr 
    mutate(
      chrm_arm = paste0(Chr, Arm), # combine the chromosome and arm into new column
      Target = target_feat # add the target feature column
    ) %>%
    select(- Arm) %>%
    arrange(desc(Gain))
    
  # Save to the complete folder
  complete_df_name <- paste0(complete_imp_folder, "/complete_", file_name)
  # write.csv(feat_joind_df, complete_df_name, row.names = FALSE, )
  
  
  # Append to the fully complete df
  complete_cancer_type_imp_df <- rbind(complete_cancer_type_imp_df, feat_joind_df)
  
    
    # Check if the feature is an aneuploidy (ends with q or p)
  if (substr(target_feat, nchar(target_feat), nchar(target_feat)) %in% c("q", "p")) {
    # Add to aneu set 
    aneu_set <- append(aneu_set, target_feat)
    # cat("\n Target_feat is an aneuploidy: ", target_feat)
    
    # ARM-LEVEL FILTERING
    # Exclude the target feature from the feature importance dataframe
    arm_lvl_filt_feat_joind_df <- feat_joind_df %>% 
      filter(chrm_arm != target_feat) 
    
    # Save the modified dataframe
    arm_lvl_filtered_df_name <- paste0(arm_lvl_filt_imp_folder, "/arm_lvl_filt_", file_name)
    # write.csv(arm_lvl_filt_feat_joind_df, arm_lvl_filtered_df_name, row.names = FALSE)
    
    # Append the fully arm level filtered df
    arm_lvl_filtered_cancer_type_imp_df <- rbind(arm_lvl_filtered_cancer_type_imp_df, arm_lvl_filt_feat_joind_df)
    
    # CHROMOSOME-LEVEL FILTERING
    # Exclude the target feature from the feature importance dataframe
    chrm_lvl_filt_feat_joind_df <- feat_joind_df %>% 
      filter(Chr != chr)
    
    # Save the modified dataframe
    chrm_lvl_filtered_df_name <- paste0(chrm_lvl_filt_imp_folder, "/chrm_lvl_filt_", file_name)
    # write.csv(chrm_lvl_filt_feat_joind_df, chrm_lvl_filtered_df_name, row.names = FALSE)
    
    # Append the fully chromosome level filtered df
    chrm_lvl_filtered_cancer_type_imp_df <- rbind(chrm_lvl_filtered_cancer_type_imp_df, chrm_lvl_filt_feat_joind_df)
    
  }
  else { # if not an aneuploidy feature
    
    # ARM-LEVEL FILTERING
    # save the file to the filtered folder
    arm_lvl_filtered_df_name <- paste0(arm_lvl_filt_imp_folder, "/filt_", file_name)
    # write.csv(feat_joind_df, arm_lvl_filtered_df_name, row.names = FALSE)
    
    # Append the fully filtered df with this target's feature importance df
    arm_lvl_filtered_cancer_type_imp_df <- rbind(arm_lvl_filtered_cancer_type_imp_df, feat_joind_df)
    
    # CHROMOSOME-LEVEL FILTERING
    # save the file to the filtered folder
    chrm_lvl_filtered_df_name <- paste0(chrm_lvl_filt_imp_folder, "/filt_", file_name)
    # write.csv(feat_joind_df, chrm_lvl_filtered_df_name, row.names = FALSE)
    
    # Append the fully filtered df with this target's feature importance df
    chrm_lvl_filtered_cancer_type_imp_df <- rbind(chrm_lvl_filtered_cancer_type_imp_df, feat_joind_df)
  }

  # # remove the variables to save memory
  objs_to_rm <- c(
    "feature_file", "file_name", "name_split", "first_part", "target_feat",
    "feature_df", "feat_joind_df", "arm_lvl_filt_feat_joind_df", 
    "chrm_lvl_filt_feat_joind_df", "arm_lvl_filtered_df_name", 
    "chrm_lvl_filtered_df_name"
  )
  rm(list = objs_to_rm[objs_to_rm %in% ls()])
  rm(objs_to_rm)
}

# Set the order for the Target features
response_features_order <- c(
  zipup(paste0(1:22, "p"), paste0(1:22, "q")) %>% unlist(),
  sort(grep("^peri", unique(complete_cancer_type_imp_df$Target), value = TRUE)),
  "ai1", "lst1", "loh_hrd"
)

# Save the full complete PANCAN feature importance df
full_complete_cancer_feat_imp_name <- paste0(feat_imp_folder, "/full_complete_PANCAN_feat_imp.csv")
# write.csv(complete_cancer_type_imp_df, full_complete_cancer_feat_imp_name, row.names = FALSE)
rm(full_complete_cancer_feat_imp_name)

# ARM-LEVEL FILTERING
# Save the full arm level filtered PANCAN feature importance df
full_arm_lvl_filtered_cancer_feat_imp_name <- paste0(feat_imp_folder, "/full_arm_lvl_filtered_PANCAN_feat_imp.csv")
# write.csv(arm_lvl_filtered_cancer_type_imp_df, full_arm_lvl_filtered_cancer_feat_imp_name, row.names = FALSE)
rm(full_arm_lvl_filtered_cancer_feat_imp_name)

# CHROMOSOME-LEVEL FILTERING
# Save the full chromosome level filtered PANCAN feature importance df
full_chrm_lvl_filtered_cancer_feat_imp_name <- paste0(feat_imp_folder, "/full_chrm_lvl_filtered_PANCAN_feat_imp.csv")
# write.csv(chrm_lvl_filtered_cancer_type_imp_df, full_chrm_lvl_filtered_cancer_feat_imp_name, row.names = FALSE)
rm(full_chrm_lvl_filtered_cancer_feat_imp_name)


# Filter the complete PANCAN feat imp df to the top genes only
top_complete_feat_imp_df <- complete_cancer_type_imp_df %>%
  group_by(Target) %>%
  slice_max(Gain, n = 1) %>%
  ungroup()

# Save the top genes feature importance for the complete df
top_complete_feat_imp_name <- paste0(feat_imp_folder, "/top_genes_complete_PACAN_feat_imp.csv")
# write.csv(top_complete_feat_imp_df, top_complete_feat_imp_name, row.names = FALSE)
rm(top_complete_feat_imp_name)

# ARM-LEVEL FILTERING
# Filter the arm level PANCAN feat imp df to the top genes only
top_arm_lvl_filtered_feat_imp_df <- arm_lvl_filtered_cancer_type_imp_df %>%
  group_by(Target) %>%
  slice_max(Gain, n = 1) %>%
  ungroup()

# Save the top genes feature importance for the df which is filtered to remove
# the genes that are cis-genes of aneuploidy features
top_arm_lvl_filtered_feat_imp_name <- paste0(feat_imp_folder, "/top_genes_filtered_PANCAN_feat_imp.csv")
# write.csv(top_arm_lvl_filtered_feat_imp_df, top_arm_lvl_filtered_feat_imp_name, row.names = FALSE)
rm(top_arm_lvl_filtered_feat_imp_name)

# CHROMOSOME-LEVEL FILTERING
# Filter the chromosome level PANCAN feat imp df to the top genes only
top_chrm_lvl_filtered_feat_imp_df <- chrm_lvl_filtered_cancer_type_imp_df %>%
  group_by(Target) %>%
  slice_max(Gain, n = 1) %>%
  ungroup()

# Save the top genes feature importance for the df which is filtered to remove
# the genes that are cis-genes of aneuploidy features
top_chrm_lvl_filtered_feat_imp_name <- paste0(feat_imp_folder, "/top_genes_chrm_lvl_filtered_PANCAN_feat_imp.csv")
# write.csv(top_chrm_lvl_filtered_feat_imp_df, top_chrm_lvl_filtered_feat_imp_name, row.names = FALSE)
rm(top_chrm_lvl_filtered_feat_imp_name)


# combine the top gene files
top_genes_combined <- top_complete_feat_imp_df %>%
  select(Target, Feature, Gain, chrm_arm) %>%
  dplyr::rename(
    Gain_complete = Gain,
    gene_loc_complt = chrm_arm,
    Feature_complt = Feature,
    Target_complt = Target
  ) %>% 
  left_join(
    top_arm_lvl_filtered_feat_imp_df %>%
      select(Target, Feature, Gain, chrm_arm) %>%
      dplyr::rename(
        Gain_filtered = Gain,
        gene_loc_arm_filt = chrm_arm,
        Feature_arm_filt = Feature,
        Target_arm_filt = Target
      ),
    by = c("Target_complt" = "Target_arm_filt")
  ) %>%
  left_join(
    top_chrm_lvl_filtered_feat_imp_df %>%
      select(Target, Feature, Gain, chrm_arm) %>%
      dplyr::rename(
        Gain_chrm_filt = Gain,
        gene_loc_chrm_filt = chrm_arm,
        Feature_chrm_filt = Feature,
        Target_chrm_filt = Target
      ),
    by = c("Target_complt" = "Target_chrm_filt")
  ) %>%
  mutate(Target_complt = factor(Target_complt, levels = response_features_order)) %>%
  arrange(Target_complt)

# Save the combined top genes feature importance
top_genes_combined_name <- paste0(feat_imp_folder, "/top_genes_combined_PANCAN_feat_imp.csv")
# write.csv(top_genes_combined, top_genes_combined_name, row.names = FALSE)
rm(top_genes_combined_name)


# Generate the plots

# For the complete data (with all genes)
# Transform the dataframe to a format suitable for heatmap
heatmap_genes_complete <- complete_cancer_type_imp_df %>%
  filter(
    Feature %in% top_complete_feat_imp_df$Feature,
  ) %>% 
  dplyr::select(Feature, Target, Gain) %>%
  pivot_wider(names_from = Feature, values_from = Gain) %>%
  replace(is.na(.), 0) %>% 
  pivot_longer(!Target, names_to = "Feature", values_to = "Gain") %>% 
  mutate(
    Target = factor(Target, levels = response_features_order),
    Gain = round(Gain, 2)
  ) %>%
  dplyr::arrange(Target, desc(Gain))

# Save the heatmap genes complete dataframe
heatmap_genes_complete_name <- paste0(feat_imp_folder, "/heatmap_genes_complete_PANCAN_feat_imp_df.csv")
# write.csv(heatmap_genes_complete, heatmap_genes_complete_name, row.names = FALSE)
rm(heatmap_genes_complete_name)

# Generate the heatmap for the complete data (with all genes)
complete_genomic_heatmap <- ggplot(heatmap_genes_complete, aes(x = Feature, y = Target, fill = Gain)) +
  geom_tile(colour = "black", lwd = 0.3) +
  scale_fill_gradient2(low = "white", high = "#D21717") +
  labs(
    title = "",
     x = "Gene",
     y = "Base Learner",
     # subtitle = paste0("PANCAN: ", cancer_type)
   ) +
  theme_minimal() +
  theme(
    axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1, color = "black", size = 10),
    axis.text.y = element_text(angle = 0, hjust = 1, color = "black", size = 11),
    plot.title = element_text(hjust = 0.5, size = 16),
    plot.subtitle = element_text( size = 12),
    legend.title = element_text(size = 15, face = "bold"),
    legend.position = "top",
    axis.title = element_text(size = 17, face = "bold"),
    plot.caption = element_text(size = 11, face = "italic")
  )

# Save the complete genomic heatmap
complete_genomic_heatmap_name <- paste0(base_feat_imp_plot, "/complete_genomic_heatmap_PANCAN.png")
# ggsave(complete_genomic_heatmap_name, plot = complete_genomic_heatmap, width = 10, height = 8, dpi = 300)
rm(complete_genomic_heatmap_name)


# ARM-LEVEL FILTERING
# Transform the dataframe to a format suitable for heatmap
heatmap_genes_arm_filt <- arm_lvl_filtered_cancer_type_imp_df %>%
  filter(
    Feature %in% top_arm_lvl_filtered_feat_imp_df$Feature,
  ) %>%
  dplyr::select(Feature, Target, Gain) %>%
  pivot_wider(names_from = Feature, values_from = Gain) %>%
  replace(is.na(.), 0) %>%
  pivot_longer(!Target, names_to = "Feature", values_to = "Gain") %>%
  mutate(
    Target = factor(Target, levels = response_features_order),
    Gain = round(as.numeric(Gain), 2)
  ) %>%
  dplyr::arrange(Target, desc(Gain))

# Save the heatmap genes arm lvl filtered dataframe
heatmap_genes_arm_filt_name <- paste0(feat_imp_folder, "/heatmap_genes_arm_filt_PANCAN_feat_imp_df.csv")
# write.csv(heatmap_genes_arm_filt, heatmap_genes_arm_filt_name, row.names = FALSE)
rm(heatmap_genes_arm_filt_name)

# Generate the heatmap for the arm lvl filtered data (with cis-genes of aneuploidy features removed)
arm_filt_genomic_heatmap <- ggplot(heatmap_genes_arm_filt, aes(x = Feature, y = Target, fill = Gain)) +
  geom_tile(colour = "black", lwd = 0.3) +
  scale_fill_gradient2(low = "white", high = "#D21717") +
  labs(
    title = "",
     x = "Gene",
     y = "Base Learner",
     # subtitle = paste0("PANCAN: ", cancer_type)
   ) +
  theme_minimal() +
  theme(
    axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1, color = "black", size = 10),
    axis.text.y = element_text(angle = 0, hjust = 1, color = "black", size = 11),
    plot.title = element_text(hjust = 0.5, size = 16),
    plot.subtitle = element_text( size = 12),
    legend.title = element_text(size = 15, face = "bold"),
    legend.position = "top",
    axis.title = element_text(size = 17, face = "bold"),
    plot.caption = element_text(size = 11, face = "italic")
  )
# Save the arm level filtered genomic heatmap
arm_filt_genomic_heatmap_name <- paste0(base_feat_imp_plot, "/arm_filt_genomic_heatmap_PANCAN.png")
# ggsave(arm_filt_genomic_heatmap_name, plot = arm_filt_genomic_heatmap, width = 10, height = 8, dpi = 300)
rm(arm_filt_genomic_heatmap_name)


# CHROMOSOME-LEVEL FILTERING
# Transform the dataframe to a format suitable for heatmap
heatmap_genes_chrm_filt <- chrm_lvl_filtered_cancer_type_imp_df %>%
  filter(
    Feature %in% top_chrm_lvl_filtered_feat_imp_df$Feature,
  ) %>%
  dplyr::select(Feature, Target, Gain) %>%
  pivot_wider(names_from = Feature, values_from = Gain) %>%
  replace(is.na(.), 0) %>%
  pivot_longer(!Target, names_to = "Feature", values_to = "Gain") %>%
  mutate(
    Target = factor(Target, levels = response_features_order),
    Gain = round(as.numeric(Gain), 2)
  ) %>%
  left_join(
    gene_annotations %>% 
      select(gene_name, chr_arm) %>% 
      dplyr::rename(Feature = gene_name, gene_loc = chr_arm),
    by = "Feature"
  ) %>% 
  mutate(
    is_gene_loc = Target == gene_loc,
    type = ifelse(
      Target %in% aneu_set, "Aneu",
      ifelse(Target %in% peri_set, "Peri", "HRD")
    )
  ) %>% 
  dplyr::arrange(Target, desc(Gain))

# Save the heatmap genes chromosome lvl filtered dataframe
heatmap_genes_chrm_filt_name <- paste0(feat_imp_folder, "/heatmap_genes_chrm_filt_PANCAN_feat_imp_df.csv")
# write.csv(heatmap_genes_chrm_filt, heatmap_genes_chrm_filt_name, row.names = FALSE)
rm(heatmap_genes_chrm_filt_name)

# Generate the heatmap for the chromosome lvl filtered data (with cis-genes of aneuploidy features removed)
chrm_filt_genomic_heatmap <- ggplot(heatmap_genes_chrm_filt, aes(x = Feature, y = Target)) +
  geom_tile(aes(fill = Gain), colour = "black", lwd = 0.3) +
  geom_tile(
    data = heatmap_genes_chrm_filt %>% filter(is_gene_loc),
    colour = "black", lwd = 0.3, fill = "black") +
  scale_fill_gradient2(low = "white", high = "#D21717") +
  labs(
    title = "",
     x = "Gene",
     y = "Base Learner",
     # subtitle = paste0("PANCAN: ", cancer_type)
   ) +
  theme_minimal() +
  theme(
    axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1, color = "black", size = 10),
    axis.text.y = element_text(angle = 0, hjust = 1, color = "black", size = 11),
    plot.title = element_text(hjust = 0.5, size = 16),
    plot.subtitle = element_text( size = 12),
    legend.title = element_text(size = 15, face = "bold"),
    legend.position = "top",
    axis.title = element_text(size = 17, face = "bold"),
    plot.caption = element_text(size = 11, face = "italic")
  ) + 
  facet_wrap(~ type, ncol = 1, scales = "free_y") 

# Save the chromosome level filtered genomic heatmap
chrm_filt_genomic_heatmap_name <- paste0(base_feat_imp_plot, "/chrm_filt_genomic_heatmap_PANCAN.png")
# ggsave(chrm_filt_genomic_heatmap_name, plot = chrm_filt_genomic_heatmap, width = 10, height = 8, dpi = 300)
rm(chrm_filt_genomic_heatmap_name)
```

```{r}
plot_2 <- ggplot(heatmap_genes_chrm_filt, aes(x = Feature, y = Target)) +
  geom_tile(aes(fill = Gain), colour = "black", lwd = 0.3) +
  # Black tiles for gene's cis-location
  geom_tile(
    data = heatmap_genes_chrm_filt %>% filter(is_gene_loc),
    colour = "black", fill = "black", lwd = 0.3
  ) +
  scale_fill_gradient2(low = "white", high = "#D21717") +
  facet_grid(type ~ ., scales = "free_y", space = "free_y") +
  theme_minimal() +
  theme(
    # x-axis text shown only on bottom panel
    axis.text.x = element_text(angle = 45, hjust = 1, size = 10),
    axis.title.x = element_text(size = 14, face = "bold"),
    axis.text.y = element_text(size = 10),
    axis.title.y = element_text(size = 14, face = "bold"),
    strip.text.y = element_text(size = 14, face = "bold"),
    legend.position = "top",
    strip.placement = "outside",
    panel.spacing.y = unit(0.9, "lines"),
    # panel.border = element_rect(colour = "black", fill = NA, size = 1),
    strip.background = element_rect(color = "black", fill = "grey88", size = 1),
  ) +
  labs(
    x = "Gene",
    y = "Target",
    fill = "Gain"
  )

print(plot_2)
```



