---
title: "Gene_Importance_modification"
output: html_document
date: "2025-06-05"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = "D:/projects/HetIN_signatures/")
```


Load the packages
```{r}
library(tidyverse)
library(ggplot2)
library(rtracklayer)
library(GenomicRanges)

```

```{r}
# Define the cancer types
can_types <- c("BRCA", "LUAD", "LUSC", "HNSC")

# folder with the feature importance files
feat_imp_folder_pre <- "data/Cancer_specific_data/Model_output"
feat_imp_folder_post <- "Results/Feature_importance/Base_feat_imp"

# Load the gene annotations 
gene_annotations <- read.csv("data/secondary_data/gene_seqnames_nnotation_df.csv", stringsAsFactors = FALSE, row.names = NULL)

# Load the gene_set genes
gene_set_genes <- read.csv("data/secondary_data/Gene_set.csv", stringsAsFactors = FALSE)
gene_set <- colnames(gene_set_genes)
rm(gene_set_genes)
```

```{r}
# reduce the gene annotations to the genes that are in the set of interest
gene_annotations <- gene_annotations %>%
  filter(gene_name %in% gene_set)
```


```{r}
for (cancer in can_types) {
  cat("\n\n\n", cancer)
  
  feat_imp_folder <- file.path(feat_imp_folder_pre, cancer, feat_imp_folder_post)
  # cat(paste0("\n\n\n", feat_imp_folder))
  
  cancer_type_imp_df <- data.frame()
  
  # Loop through each feature importance files in the folder
  for (feature_file in list.files(feat_imp_folder, full.names = TRUE)) {
      # cat("\n\n", feature_file)
      
      # Get the feature from the name
      file_name <- basename(feature_file)
      target_feat <- str_split(file_name, "_")[[1]][1]
      # cat("\n\n Target_feat: ", target_feat)
      
      # Check if the feature is an aneuploidy (ends with q or p)
      if (substr(target_feat, nchar(target_feat), nchar(target_feat)) %in% c("q", "p")) {
        cat("\n Target_feat is an aneuploidy: ", target_feat)
        
        feature_chrom <-gsub("[^0-9]", "", target_feat) %>% as.character()
        feature_arm <- substr(target_feat, nchar(target_feat), nchar(target_feat))
        cat("\n Chromosome: ", feature_chrom)
        
        # Read the CSV file
        feature_df <- read.csv(feature_file, stringsAsFactors = FALSE, row.names = NULL) %>% 
          select(-X)
        
        # Join the dfs
        feat_joind_df <- feature_df %>%
          left_join(
            gene_annotations %>% 
              select(c("gene_name", "seqnames", "arm")), 
            by = c("Feature" = "gene_name")
            ) %>%
          rename(
            "Chr" = "seqnames",
            "Arm" = "arm"
          ) %>% 
          mutate(
            chrm_arm = paste0(Chr, Arm)
          )
        
        # Resave the originial with the chr col
        ori_df_name <- paste0(feat_imp_folder, "/ori_", file_name)
        # write.csv(feat_joind_df, ori_df_name, row.names = FALSE)
        
        # Filter the DF
        filt_feat_joind_df <- feat_joind_df %>% 
          filter(
            chrm_arm != target_feat,
            is.na(Chr) == FALSE,
            is.na(Arm) == FALSE
          )
        
        # Save the modified dataframe
        df_name <- paste0(feat_imp_folder, "/mod_", file_name)
        # cat("\n", df_name)
        # write.csv(filt_feat_joind_df, df_name, row.names = FALSE)
        
        # Add target feature col and append to full df
        mod_feat_df <- filt_feat_joind_df %>% 
          mutate(
            "Target" = target_feat
          )
        
        cancer_type_imp_df <- rbind(cancer_type_imp_df, mod_feat_df)
        
      } 
      else if (target_feat %in% c("mod","ori")) {
        next
      }
      else {
        cat("\n Feature is not an aneuploidy: ", target_feat)
      }
  }
}
```




